controller/object_store_controller.go

```diff
@@ -0,0 +1,1139 @@
+// create them.
+// To manage their interaction and ensure cleanup when the object store is
+// deleted, K8s owner ship relations (aka. owner references) are used with the
+// following relations between the K8s objects:
+//
+// | ┌────────────────┐
+// | │ ObjectStore    │
+// | └─┬──────────────┘
+// |   │
+// |   │owns
+// |   │
+// |   │     ┌───────────────────────┐
+// |   ├────►│ Service               │
+// |   │     └───────────────────────┘
+// |   │
+// |   │     ┌───────────────────────┐
+// |   ├────►│ UI Ingress            │
+// |   │     └───────────────────────┘
+// |   │
+// |   │     ┌───────────────────────┐
+// |   ├────►│ optional S3 Ingresses │
+// |   │     └───────────────────────┘
+// |   │
+// |   │     ┌───────────────────────┐
+// |   ├────►│ Deployment            ├──────┐
+// |   │     └───────────────────────┘      │owns
+// |   │                                    ▼
+// |   │     ┌───────────────────────┐    ┌────────────────┐
+// |   └────►│ PersistentVolumeClaim │    │ ReplicaSet     │
+// |         └─┬─────────────────────┘    └─┬──────────────┘
+// |           │owns                        │owns
+// |           │                            │
+// |           ▼                            ▼
+// |         ┌───────────────────────┐    ┌────────────────┐
+// |         │ LonghornVolume        │    │ Pod            │
+// |         └───────────────────────┘    └────────────────┘
+// |                                        ▲
+// |                                        │
+// |                                        │waits for
+// |         ┌───────────────────────┐      │shutdown
+// |         │ PersistentVolume      ├──────┘
+// |         └───────────────────────┘
+//
+// From this ownership relationship and the mount dependencies, the order of
+// creation of the resources is determined.
+func (osc *ObjectStoreController) handleStarting(store *longhorn.ObjectStore) (err error) {
+	pvc, store, err := osc.getOrCreatePVC(store)
+	if err != nil {
+		return errors.Wrap(err, "API error while creating pvc")
+	}
+
+	vol, store, err := osc.getOrCreateVolume(store, pvc)
+	if err != nil {
+		return errors.Wrap(err, "API error while creating volume")
+	}
+
+	pv, store, err := osc.getOrCreatePV(store, vol)
+	if err != nil {
+		return errors.Wrap(err, "API error while creating volume")
+	}
+
+	dpl, store, err := osc.getOrCreateDeployment(store)
+	if err != nil {
+		return errors.Wrap(err, "API error while creating deployment")
+	}
+
+	svc, store, err := osc.getOrCreateService(store)
+	if err != nil {
+		return errors.Wrap(err, "API error while creating service")
+	}
+
+	ingress, store, err := osc.getOrCreateIngress(store)
+	if err != nil {
+		return errors.Wrap(err, "API error while creating UI ingress")
+	}
+
+	endpoints, store, err := osc.getOrCreateS3Endpoints(store)
+	if err != nil {
+		return errors.Wrap(err, "API error while creating S3 ingresses")
+	}
+	osc.logger.Info("created %d S3 endpoint(s)", len(endpoints))
+
```

Comment:
Should we set the state `store.Status.State = longhorn.ObjectStoreStateStarting` because if we reach this line, all resource are created and we are waiting for them to be running?